cirq-core/cirq/transformers/optimize_for_target_gateset.py

```diff
@@ -126,20 +132,38 @@ def optimize_for_target_gateset(
         return _decompose_operations_to_target_gateset(
             circuit, context=context, ignore_failures=ignore_failures
         )
+    if isinstance(max_num_passes, int):
+        _outerloop = lambda: range(max_num_passes)
+    else:
+
+        def _outerloop():
+            while True:
+                yield 0
+
+    initial_num_moments, initial_num_ops = len(circuit), len(tuple(circuit.all_operations()))
+    for _ in _outerloop():
+        for transformer in gateset.preprocess_transformers:
+            circuit = transformer(circuit, context=context)
+
+        circuit = _decompose_operations_to_target_gateset(
+            circuit,
+            context=context,
+            gateset=gateset,
+            decomposer=gateset.decompose_to_target_gateset,
+            ignore_failures=ignore_failures,
+            tags_to_decompose=(gateset._intermediate_result_tag,),
+        )
 
-    for transformer in gateset.preprocess_transformers:
-        circuit = transformer(circuit, context=context)
-
-    circuit = _decompose_operations_to_target_gateset(
-        circuit,
-        context=context,
-        gateset=gateset,
-        decomposer=gateset.decompose_to_target_gateset,
-        ignore_failures=ignore_failures,
-        tags_to_decompose=(gateset._intermediate_result_tag,),
-    )
+        for transformer in gateset.postprocess_transformers:
+            circuit = transformer(circuit, context=context)
 
-    for transformer in gateset.postprocess_transformers:
-        circuit = transformer(circuit, context=context)
+        circuit = circuits.Circuit(
+            op for op in circuit.all_operations()
+        )  # Ensure the circuit is contracted.
```

Comment:
This leads to a loss of moment structure. When we wrote this function originally, preserving moment structure was an important requirement. Are you sure we want to do this?

Reply:
what was the original requirement exactly? the circuit in the issue is supposed to endup with 3 moments as per @eliottrosenberg which can't happen without some sort of contraction. probably there is a way to do the contraction without destroying the moment structure. but we need to clarify what it's exactly that we want to preserve or if we simply need another `preserve_moment_structure` parameter

Reply:
Yeah, I think reducing the number of moments is generally a good thing. In general, this can't preserve the number of moments (i.e. if I ask it to transform a `cirq.ISWAP` gate into CZ).

Reply:
@tanujkhattar apart from this line. the current implementation doesn't preserve the moment structure, for example look at the output in the issue. CZ(4, 5) is earlier than it should and PhXZ(a=0,x=1,z=0)(6)───PhXZ(a=0.5,x=-0.5,z=1)(6)-CZ(5, 6) are a lot later than they should.

Reply:
Let's continue the discussion on the original issue here - https://github.com/quantumlib/Cirq/issues/6422#issuecomment-1912760153